src/main/java/org/jenkinsci/plugins/mesos/MesosApi.java

```diff
@@ -49,18 +69,57 @@ public MesosApi(String masterUrl, String user, String frameworkName)
             .withValue("master-url", ConfigValueFactory.fromAnyRef(masterUrl));
     this.clientSettings = MesosClientSettings.fromConfig(conf);
     system = ActorSystem.create("mesos-scheduler");
+    context = system.dispatcher();
     materializer = ActorMaterializer.create(system);
 
     client = connectClient().get();
+
+    stateMap = new ConcurrentHashMap<>();
+
+    logger.info("Starting USI scheduler flow.");
+    updates = runUsi(SpecsSnapshot.empty(), client, materializer);
+  }
+
+  /**
+   * Constructs a queue of {@link SpecUpdated} and passes the specs snapshot as the first item. All
+   * updates are processed by {@link MesosApi#updateState(StateEvent)}.
+   *
+   * @param specsSnapshot The initial set of pod specs.
+   * @param client The Mesos client that is used.
+   * @param materializer The {@link ActorMaterializer} used for the source queue.
+   * @return A running source queue.
+   */
+  private SourceQueueWithComplete<SpecUpdated> runUsi(
+      SpecsSnapshot specsSnapshot, MesosClient client, ActorMaterializer materializer) {
+    var schedulerFlow = Scheduler.fromSnapshot(specsSnapshot, client);
+
+    // We create a SourceQueue and assume that the very first item is a spec snapshot.
+    var queue =
+        Source.<SpecUpdated>queue(256, OverflowStrategy.fail())
+            .via(schedulerFlow)
+            .flatMapConcat(pair -> pair.second()) // Ignore state snapshot for now.
```

Comment:
What if the scheduler fails? flatMapConcat will not propagate any failures from the downstream. Without proper error handling this code is not safe.

Reply:
What do you mean? If the scheduler has an exception? I believe the scheduler flow should handle this, propagate the error and close the stream. No user should deal with that.

Reply:
The problem that the error won't be propagated because of flatMapConcat. Why don't you want to use my adapter which already handles all those edge cases?